subsys/zigbee/cli/zigbee_cli_cmd_zdo.c

```diff
@@ -0,0 +1,2126 @@
+	"Usage: mgmt_leave <h:16-bit dst_addr> [h:device_address eui64] " \
+		"[--children] [--rejoin]\n" \
+	"--children - Device should also remove its children when leaving.\n" \
+	"--rejoin - Device should rejoin network after leave.")
+
+#define MGMT_LQI_HELP \
+	("Perform mgmt_lqi request.\n" \
+	"Usage: mgmt_lqi <h:short> [d:start index]")
+
+/* Defines how many ZDO requests can be run concurrently. */
+#define ZIGBEE_CLI_ZDO_TSN                 3
+/* Defines how long to wait, in seconds, for Match Descriptor Response. */
+#define ZIGBEE_CLI_MATCH_DESC_RESP_TIMEOUT 5
+/* Defines how long to wait, in seconds, for Bind Response. */
+#define ZIGBEE_CLI_BIND_RESP_TIMEOUT       5
+/* Defines how long to wait, in seconds, for Network Addrees Response. */
+#define ZIGBEE_CLI_NWK_ADDR_RESP_TIMEOUT   5
+/* Defines how long to wait, in seconds, for IEEE (EUI64) Addrees Response. */
+#define ZIGBEE_CLI_IEEE_ADDR_RESP_TIMEOUT  5
+/* Defines how long to wait, in seconds, for mgmt_leave response. */
+#define ZIGBEE_CLI_MGMT_LEAVE_RESP_TIMEOUT 5
+
+
+LOG_MODULE_DECLARE(cli);
+
+typedef struct {
+	/* Starting Index for the requested elements. */
+	zb_uint8_t  start_index;
+	/* Destination address. */
+	zb_uint16_t dst_addr;
+} req_seq_t;
+
+/* Structure to store inforamtion required for sending zdo requests.  */
+typedef struct {
+	zb_bufid_t         buffer_id;
+	zb_uint8_t         ctx_timeout;
+	zb_uint8_t         (*req_fn)(zb_uint8_t param, zb_callback_t cb_fn);
+	void               (*timeout_cb_fn)(zb_bufid_t bufid);
+	void               (*req_cb_fn)(zb_bufid_t bufid);
+} zdo_req_info_t;
+
+/* Forward declarations. */
+static void ctx_timeout_cb(zb_uint8_t tsn);
+
+/* This structure allows for binding ZBOSS transaction and CLI object.. */
+typedef struct zdo_tsn_ctx {
+	const struct shell *shell;
+	bool               (*p_cb_fn)(struct zdo_tsn_ctx *zdo_tsn_ctx, u8_t param);
+	u8_t               tsn;
+	bool               is_broadcast;
+	atomic_t           taken;
+	union {
+		/* Extra context for commands which request tables. */
+		req_seq_t  req_seq;
+	} cmd_ctx;
+	zdo_req_info_t     zdo_req;
+} zdo_tsn_ctx_t;
+
+static zdo_tsn_ctx_t m_tsn_ctx[ZIGBEE_CLI_ZDO_TSN];
+
+/**@brief Return a pointer to context with the given transaction
+ *        sequence number.
+ *
+ * @param[in] tsn ZBOSS transaction sequence number.
+ *
+ * @return a pointer to context or NULL if context for given TSN wasn't found.
+ */
+static zdo_tsn_ctx_t *get_ctx_by_tsn(u8_t tsn)
+{
+	for (u8_t i = 0; i < ARRAY_SIZE(m_tsn_ctx); i++) {
+		if ((atomic_get(&m_tsn_ctx[i].taken)) &&
+		    (m_tsn_ctx[i].tsn == tsn)) {
+			return &m_tsn_ctx[i];
+		}
+	}
+
+	return NULL;
+}
+
+/**@brief Get a pointer to a free context.
+ *
+ * @return a pointer to context structure or NULL if all contexts are taken.
+ */
+static zdo_tsn_ctx_t *get_free_ctx(void)
+{
+	for (u8_t i = 0; i < ARRAY_SIZE(m_tsn_ctx); i++) {
+		if (!atomic_get(&m_tsn_ctx[i].taken)) {
+			atomic_set(&m_tsn_ctx[i].taken, true);
```

Comment:
The usage of atomic type in this PR is a bit questionable. Who do you ignore return values?